### ISSUE TITLE
support for aten::embedding + quick question

### ISSUE THREAD

**Issue by joechen018** [aws-neuron/aws-neuron-sdk#580]
Hi! I tried to run neuron.trace on a whisper model, and it compiled, however it was slower and inaccurate. Even though it compiled, not all operations were supported (see below).

I noticed that aten::embedding was removed in release [1.0.1001.0](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/release-notes/compiler/neuron-cc/neuron-cc-ops/neuron-cc-ops-pytorch.html#pytorch-neuron-release-1-0-1001-0).

Just had a few quick question:
1) If 99.9% of operations compiled, as in my case, should I expect the accuracy to be roughly identical to the original model, or is it at risk of significant error? I'm trying to see whether the poor performance is due to the the non-supported operation or due an implementation error on my end.
2) If it is due to the non-supported operation, is there anything I can do to get around it? Or will there be support for this in the future?

```
import torch
import torch.neuron
import whisper

whisper_model = whisper.load_model("small")

loaded_small = torch.load('example_small.pt')
x_small = loaded_small['token']
xa_small = loaded_small['audio']
out_reg = whisper_model.decoder(x_small, xa_small)

torch.neuron.analyze_model(whisper_model.decoder, tuple([x_small,xa_small]))
whisper_decoder = torch.neuron.trace(whisper_model.decoder, example_inputs=tuple([x_small,xa_small]))
```
This file is referenced in script above in case you want to replicate.
[example_small.pt.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/9825127/example_small.pt.zip)

Below is the output of torch.neuron.trace

```
INFO:Neuron:There are 1 ops of 1 different types in the TorchScript that are not compiled by neuron-cc: aten::embedding, (For more information see https://github.com/aws/aws-neuron-sdk/blob/master/release-notes/neuron-cc-ops/neuron-cc-ops-pytorch.md)
INFO:Neuron:Number of arithmetic operators (pre-compilation) before = 442, fused = 441, percent fused = 99.77%
INFO:Neuron:Number of neuron graph operations 1565 did not match traced graph 1552 - using heuristic matching of hierarchical information
WARNING:tensorflow:From /home/ec2-user/pytorch_venv/lib64/python3.7/site-packages/torch_neuron/ops/aten.py:2022: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
INFO:Neuron:Compiling function _NeuronGraph$462 with neuron-cc
INFO:Neuron:Compiling with command line: '/home/ec2-user/pytorch_venv/bin/neuron-cc compile /tmp/tmpmwf5k9yv/graph_def.pb --framework TENSORFLOW --pipeline compile SaveTemps --output /tmp/tmpmwf5k9yv/graph_def.neff --io-config {"inputs": {"0:0": [[1, 1], "int64"], "1:0": [[1, 1, 384], "float32"], "2:0": [[1, 1500, 384], "float32"]}, "outputs": ["aten_matmul/MatMul:0"]} --verbose 35'
....
Compiler status PASS
INFO:Neuron:skip_inference_context for tensorboard symbols at /home/ec2-user/pytorch_venv/lib64/python3.7/site-packages/torch_neuron/tensorboard.py:305 tb_parse
INFO:Neuron:Number of neuron graph operations 1565 did not match traced graph 1552 - using heuristic matching of hierarchical information
INFO:Neuron:Number of arithmetic operators (post-compilation) before = 442, compiled = 441, percent compiled = 99.77%
INFO:Neuron:The neuron partitioner created 1 sub-graphs
INFO:Neuron:Neuron successfully compiled 1 sub-graphs, Total fused subgraphs = 1, Percent of model sub-graphs successfully compiled = 100.0%
INFO:Neuron:Compiled these operators (and operator counts) to Neuron:
INFO:Neuron: => aten::Int: 57
INFO:Neuron: => aten::add: 18
INFO:Neuron: => aten::div: 8
INFO:Neuron: => aten::flatten: 8
INFO:Neuron: => aten::gelu: 4
INFO:Neuron: => aten::layer_norm: 13
INFO:Neuron: => aten::linear: 40
INFO:Neuron: => aten::matmul: 17
INFO:Neuron: => aten::mul: 16
INFO:Neuron: => aten::permute: 32
INFO:Neuron: => aten::pow: 8
INFO:Neuron: => aten::size: 61
INFO:Neuron: => aten::slice: 9
INFO:Neuron: => aten::softmax: 8
INFO:Neuron: => aten::to: 117
INFO:Neuron: => aten::transpose: 1
INFO:Neuron: => aten::view: 24
INFO:Neuron:Not compiled operators (and operator counts) to Neuron:
INFO:Neuron: => aten::embedding: 1 [not supported]
INFO:Neuron:skip_inference_context for tensorboard symbols at /home/ec2-user/pytorch_venv/lib64/python3.7/site-packages/torch_neuron/tensorboard.py:305 tb_parse
INFO:Neuron:Number of neuron graph operations 6 did not match traced graph 9 - using heuristic matching of hierarchical information
```

**Comment by awsilya** [comment]
Hi Joe,

Unsupported operators are executed on CPU, which means they should not degrade accuracy at all.

The embedding operation will likely not cause a performance issue here because it appears it is at the very beginning of the graph. This means that only one operation is executed on CPU and the remaining large subgraph is executed on Neuron. Unsupported operators most often cause issues when they are scattered throughout a model and cause the graph to be partitioned into many pieces. Note in the logs that the compilation process produced 1 subgraph, which is the ideal case.

It is unlikely that we will enable embeddings by default on inf1, but this most likely is not causing a performance problem here.

Generally accuracy issues are caused by performance optimizations that can be tuned according to the model requirements. The best place to start is to see if you can achieve the accuracy needed by disabling fast math:
whisper_decoder = torch.neuron.trace(whisper_model.decoder, example_inputs=tuple([x_small,xa_small]), compiler_args='--fast-math=none')

For more details, see the mixed precision guide here: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/neuron-cc/mixed-precision.html#neuron-cc-training-mixed-precision
One potential solution to address performance is to ensure that all NeuronCores are in use. Unlike other accelerators, the minimum number of accelerators on an inf1 instance is 4 NeuronCores. This means that generally to maximize performance you want  4 models loaded. Alternatively there are tools like DataParallel that could help for batched requests: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/torch-neuron/torch-neuron-dataparallel-app-note.html

**Comment by joechen018** [comment]
Thanks -- this is very helpful. I still need to solve the accuracy issue and may have follow-up questions. In the meantime, I was wondering:

1) When I tried running the neuron model, I get warnings that look like
```
[W runtime.cpp:370] Warning: Unable to find optimal NeuronCore placement for a model with 25 cores. (function get_core_group)
```
Does this mean I need an instance with 25 cores? Does this mean I'm not properly using all the cores at once?

2) My use case is for real time transcription, and I want the most cost-efficient model to handle requests as fast as possible. If I ran an optimized neuron (like below), would that only be using one core? If so, would I be able to leverage multiple cores to have linear speedup for a single request or does using multiple cores only allow for an increase in throughput via parallel processing? I'm wondering because each time I switch to a larger inf instance to compute my model (this is a before I compiled it on neuron), there would be a speed-up -- and I wasn't sure if that meant more cores = speedup or if only one core is responsible for a model at a given time and larger instances just have "better" cores. This might just be a lack of computer architecture knowledge on my end, but could you help clarify my understanding?
```
model_neuron = torch.jit.load('whisper_decoder.pt')
model_neuron(x, xa).shape
```

**Comment by fareesh** [comment]
I'm in a similar situation (although not looking at solving real-time inference for the moment), and I'm using the medium model.

I'm not too familiar with neuron but I'm keen to try it out for the advertised cost savings.

I'm currently in the middle of compiling/tracing and it's taking a while (probably 20-30 minutes so far). Since there's some trial and error suggested here (disabling fast math, etc.), would this be faster if I launched an instance with a different configuration? (I'm on inf1.xlarge at present). Any other ways to speed this up?

**Comment by joechen018** [comment]
@fareesh 
I was looking at a few tutorials and they suggested compiling on inf1.6xlarge [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/resnet50.html). I compiled on a 2x large c6 instance and then loaded it on a inf1.xlarge afterwards.

**Comment by fareesh** [comment]
@joechen018 thanks I will give that a try!

**Comment by fareesh** [comment]
I can't seem to import the model after it compiles. Here's the output I encounter

```
2022-Oct-23 06:47:23.0297  1413:1413  ERROR  NMGR:get_ops_info                            Get Neuron Ops error: [06:47:23] /local/p4clients/pkgbuild-oj2lE/workspace/src/KaenaRuntime/kmgr/tvm/dmlc/json.h:695: unknown string escape \u
Stack trace:
  [bt] (0) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(+0x1302dd) [0x7f334d5e42dd]
  [bt] (1) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(+0x13700d) [0x7f334d5eb00d]
  [bt] (2) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(+0x13c178) [0x7f334d5f0178]
  [bt] (3) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(+0x134bc1) [0x7f334d5e8bc1]
  [bt] (4) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(+0x134f5a) [0x7f334d5e8f5a]
  [bt] (5) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(+0x12c855) [0x7f334d5e0855]
  [bt] (6) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(+0x12e7d9) [0x7f334d5e27d9]
  [bt] (7) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libnrt.so(nrt_load+0xc7) [0x7f334d4fb727]
  [bt] (8) /home/ubuntu/pytorch_venv/lib/python3.7/site-packages/torch_neuron/lib/libtorchneuron.so(neuron::NeuronModel::NeuronModel(std::string const&, std::basic_string_view<char, std::char_traits<char> > const&, int, int)+0x5fb) [0x7f334cffba3b]


2022-Oct-23 06:47:23.0298  1413:1413  ERROR  NMGR:kmgr_load_nn_post_metrics               Failed to load NN: 1.11.7.0+aec18907e-/tmp/tmpfpb0r5an, err: 2
```

**Comment by awsilya** [comment]
@fareesh this ^^^ is clearly a bug, we have not see it before.  Could you please open a new issue?  Thanks

**Comment by awsilya** [comment]
re: compiling.  Yes, we advise to compile on C6 because of better price/performance.  Since compilation does not use ML acceleration hardware, there is no reason to pay for it during compilation.  

inf1.xl has too few VCPUs which might increase the compilation time.  Use at least 6xl.

**Comment by awsilya** [comment]
[W runtime.cpp:370] Warning: Unable to find optimal NeuronCore placement for a model with 25 cores 

This means the compiled graph requires 25 cores to run efficiently.  1/2xl have only 4 cores which is why we show the warning.  The only instance type that has >25 cores is 24xl.

BUT there is a decent chance you'll not be happy with price/performance.  Unless compiled for neuron pipeline execution, presumably not in this case, large number of NCs means the graph was split into multiple subgraphs during compilation.  That is caused by the presence of unsupported operators in the graph.  If you would be willing to share your compiled artifacts we can take closer look.

**Comment by awsilya** [comment]
cores/price/performance - we need to sort out compilation & operator support problem first.  If the model is compiled to require 25  NC because of graph fragmentation, the chances of getting good results are pretty low.

Assume we can sort this out and compile for a single NC.  Than depending on the instance size you can run multiple inferences in parallel.  For example if you choose to run on 2xl you could execute 4 inferences in parallel, on 6xl you could execute 16 inferences in parallel etc.

**Comment by jluntamazon** [comment]
We were able to reproduce this issue, we have identified the problem, and we are looking at implementing a fix

**Comment by mrnikwaws** [comment]
We are continuing to work on this issue and expect to deliver the fix in an upcoming release

**Comment by awsrjh** [comment]
Hi

we released a fix for this in our latest release: 

https://awsdocs-neuron.readthedocs-hosted.com/en/latest/release-notes/index.html#neuron-2-5-0-11-23-2022
